### PR TITLE
SDL-0255 Enhance BodyInformation vehicle data

### LINKED ISSUE #1477: [SDL 0255] Enhance BodyInformation vehicle data
Proposal: [Enhance BodyInformation vehicle data](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0255-Enhance-BodyInformation-vehicle-data.md)

> This proposal is to enhance BodyInformation with more params to get access to more door related signals.

Review: https://github.com/smartdevicelink/sdl_evolution/issues/847

### PR DESCRIPTION
**Pull request by yoooriii**
Fixes #1477 

This PR is **ready** for review.

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior

#### Unit Tests
Added 3 new unit tests for 3 new structs
* SDLDoorStatusSpec
* SDLGateStatusSpec
* SDLRoofStatusSpec

Added a new unit test for new enum
* SDLDoorStatusTypeSpec

Updated old unit test
* SDLBodyInformationSpec

#### Core Tests
[List of tests performed against Core and behaviors verified]

Core version / branch / commit hash / module tested against: [Core: sdl_0255_enhance_bodyInformation_vehicle_data](https://github.com/LuxoftSDL/sdl_core/tree/feature/sdl_0255_enhance_bodyInformation_vehicle_data)
HMI name / version / branch / commit hash / module tested against: [HMI: sdl_0255_enhance_bodyInformation_vehicle_data](https://github.com/LuxoftSDL/sdl_hmi/tree/feature/sdl_0255_enhance_bodyInformation_vehicle_data)

### Summary
Added 3 new structs
* SDLDoorStatus
* SDLGateStatus
* SDLRoofStatus

Added a new enum
* SDLDoorStatusType

Updated an old struct
* SDLBodyInformationSpec

##### Breaking Changes
* no breaking changes

##### Enhancements
* no enhancements

##### Bug Fixes
* no bug fixes

### Tasks Remaining:
- [x] none

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by yoooriii** [comment]
Hello @atiwari9. Could you please review the PR? Thank you.

**Comment by atiwari9** [comment]
@SatbirTanda  - Please take note..

**Comment by yoooriii** [comment]
Hello @atiwari9, @mrapitis. Could you please review the PR? Thank you.

**Comment by atiwari9** [comment]
@standa1-ford  - Please take note.

File: SmartDeviceLink-iOS.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1663,6 +1663,18 @@
...
+		B3838A21257C5BB000420C11 /* SDLRoofStatus.m in Sources */ = {isa = PBXBuildFile; fileRef = B3838A1F257C5BB000420C11 /* SDLRoofStatus.m */; };
+		B3838A28257C5CE600420C11 /* SDLDoorStatusType.h in Headers */ = {isa = PBXBuildFile; fileRef = B3838A26257C5CE600420C11 /* SDLDoorStatusType.h */; settings = {ATTRIBUTES = (Public, ); }; };
+		B3838A29257C5CE600420C11 /* SDLDoorStatusType.m in Sources */ = {isa = PBXBuildFile; fileRef = B3838A27257C5CE600420C11 /* SDLDoorStatusType.m */; };
+		B3838A30257C5D1B00420C11 /* SDLGateStatus.m in Sources */ = {isa = PBXBuildFile; fileRef = B3838A2E257C5D1A00420C11 /* SDLGateStatus.m */; };
+		B3838A31257C5D1B00420C11 /* SDLGateStatus.h in Headers */ = {isa = PBXBuildFile; fileRef = B3838A2F257C5D1B00420C11 /* SDLGateStatus.h */; settings = {ATTRIBUTES = (Public, ); }; };
+		B3838A3A257C6AB700420C11 /* SDLDoorStatus.m in Sources */ = {isa = PBXBuildFile; fileRef = B3838A38257C6AB600420C11 /* SDLDoorStatus.m */; };
+		B3838A3B257C6AB700420C11 /* SDLDoorStatus.h in Headers */ = {isa = PBXBuildFile; fileRef = B3838A39257C6AB600420C11 /* SDLDoorStatus.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Comment (standa1-ford) [inline SmartDeviceLink-iOS.xcodeproj/project.pbxproj:1677]: please add files in proper alphabetical order

**Review by standa1-ford — APPROVED** [review]
Great work!!

**Comment by NicoleYarroch** [comment]
@yoooriii Can you tag me when you are ready to have Livio review the PR? Thanks!

**Comment by yoooriii** [comment]
Hello @jordynmackool. Could you review please?

**Comment by a-prykho** [comment]
Hi @NicoleYarroch , this PR is ready for Livio review.

**Review by NicoleYarroch — CHANGES_REQUESTED** [review]
I found a few minor things to fix in the code base. Also, the test cases need to be refactored so they work correctly with Xcode's code coverage.

**Comment by NicoleYarroch** [comment]
@yoooriii Please tag me when you are ready for a re-review. Thanks.

**Comment by yoooriii** [comment]
> @yoooriii Please tag me when you are ready for a re-review. Thanks.

Hello @NicoleYarroch, I have updated the PR.  Could you please re-review my changes? Thank you.
(for some reason tests on server fail with an error: *'Quick/Quick.h' file not found*. Which is not related to this PR)

File: SmartDeviceLink/public/SDLBodyInformation.m [hunk 2]
```diff
@@ -17,23 +17,21 @@ @implementation SDLBodyInformation
 
 - (instancetype)initWithParkBrakeActive:(BOOL)parkBrakeActive ignitionStableStatus:(SDLIgnitionStableStatus)ignitionStableStatus ignitionStatus:(SDLIgnitionStatus)ignitionStatus {
     self = [self init];
-    if (!self) {
-        return nil;
+    if (self) {
+        self.parkBrakeActive = @(parkBrakeActive);
+        self.ignitionStableStatus = ignitionStableStatus;
+        self.ignitionStatus = ignitionStatus;
```
Comment (NicoleYarroch) [inline SmartDeviceLink/public/SDLBodyInformation.m:23]: Can you change this back to the way the generator created it? Then the formatting will be consistent with most of the other classes in the project.

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLBodyInformationSpec.m [hunk 3]
```diff
@@ -9,62 +9,183 @@
...
+NSArray<SDLDoorStatus *> *doorStatuses = @[[[SDLDoorStatus alloc] init]];
+NSArray<SDLGateStatus *> *gateStatuses = @[[[SDLGateStatus alloc] init]];
+NSArray<SDLRoofStatus *> *roofStatuses = @[[[SDLRoofStatus alloc] init]];
+SDLIgnitionStableStatus ignitionStableStatus = SDLIgnitionStableStatusStable;
+SDLIgnitionStatus ignitionStatus = SDLIgnitionStatusStart;
+__block SDLBodyInformation* testStruct = nil;
+
+describe(@"getter/setter tests", ^{
+    afterEach(^{
+        testStruct = nil;
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLBodyInformationSpec.m:30]: This `afterEach(^{})` is unnecessary and can be removed.

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLBodyInformationSpec.m [hunk 4]
```diff
@@ -9,62 +9,183 @@
...
-        expect(testStruct.driverDoorAjar).to(equal(@NO));
-        expect(testStruct.passengerDoorAjar).to(equal(@NO));
-        expect(testStruct.rearLeftDoorAjar).to(equal(@NO));
-        expect(testStruct.rearRightDoorAjar).to(equal(@YES));
+NSArray<SDLDoorStatus *> *doorStatuses = @[[[SDLDoorStatus alloc] init]];
+NSArray<SDLGateStatus *> *gateStatuses = @[[[SDLGateStatus alloc] init]];
+NSArray<SDLRoofStatus *> *roofStatuses = @[[[SDLRoofStatus alloc] init]];
+SDLIgnitionStableStatus ignitionStableStatus = SDLIgnitionStableStatusStable;
+SDLIgnitionStatus ignitionStatus = SDLIgnitionStatusStart;
+__block SDLBodyInformation* testStruct = nil;
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLBodyInformationSpec.m:26]:
```suggestion
__block SDLBodyInformation *testStruct = nil;
```

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLBodyInformationSpec.m [hunk 5]
```diff
@@ -9,62 +9,183 @@
...
-        expect(testStruct.ignitionStableStatus).to(beNil());
-        expect(testStruct.ignitionStatus).to(beNil());
-        expect(testStruct.driverDoorAjar).to(beNil());
-        expect(testStruct.passengerDoorAjar).to(beNil());
-        expect(testStruct.rearLeftDoorAjar).to(beNil());
-        expect(testStruct.rearRightDoorAjar).to(beNil());
+
+    context(@"initWithDictionary:", ^{
+        beforeEach(^{
+            NSDictionary* dict = @{
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLBodyInformationSpec.m:74]:
```suggestion
            NSDictionary *dict = @{
```

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLDoorStatusSpec.m [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+#import "SDLDoorStatus.h"
+#import "SDLDoorStatusType.h"
+#import "SDLGrid.h"
+#import "SDLRPCParameterNames.h"
+
+QuickSpecBegin(SDLDoorStatusSpec)
+
+SDLGrid *location = [[SDLGrid alloc] init];
+SDLDoorStatusType status = SDLDoorStatusTypeAjar;
+__block SDLDoorStatus* testStruct = nil;
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLDoorStatusSpec.m:20]:
```suggestion
__block SDLDoorStatus *testStruct = nil;
```

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLDoorStatusSpec.m [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+describe(@"getter/setter tests", ^{
+    afterEach(^{
+        testStruct = nil;
+    });
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLDoorStatusSpec.m:25]: This `afterEach(^{})` is necessary and can be removed.

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLDoorStatusSpec.m [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        beforeEach(^{
+            testStruct = [[SDLDoorStatus alloc] initWithLocation:location status:status];
+        });
+
+        it(@"expect struct is not nil", ^{
+            expect(testStruct).notTo(beNil());
+        });
+
+        testStruct.location = location;
+        testStruct.status = status;
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLDoorStatusSpec.m:88]:
```
testStruct.location = location;
testStruct.status = status;
```
Needs to be inside an `it(@"", ^{})`

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLDoorStatusSpec.m [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+        it(@"expect all properties to be set correctly", ^{
+            expect(testStruct.location).to(equal(location));
+            expect(testStruct.status).to(equal(status));
+        });
+    });
+
+    context(@"initWithDictionary:", ^{
+        beforeEach(^{
+            NSDictionary* dict = @{
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLDoorStatusSpec.m:46]:
```suggestion
            NSDictionary *dict = @{
```

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLGateStatusSpec.m [hunk 10]
```diff
@@ -0,0 +1,97 @@
+//
+//  SDLGateStatus.m
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLGateStatusSpec.m:2]:
```suggestion
//  SDLGateStatusSpec.m
```

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLGateStatusSpec.m [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+QuickSpecBegin(SDLGateStatusSpec)
+
+SDLGrid *location = [[SDLGrid alloc] init];
+SDLDoorStatusType status = SDLDoorStatusTypeAjar;
+__block SDLGateStatus *testStruct = nil;
+
+describe(@"getter/setter tests", ^{
+    afterEach(^{
+        testStruct = nil;
+    });
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLGateStatusSpec.m:26]: This `afterEach(^{})` is necessary and can be removed.

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLGateStatusSpec.m [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+        it(@"expect all properties to be set correctly", ^{
+            expect(testStruct.location).to(equal(location));
+            expect(testStruct.status).to(equal(status));
+        });
+    });
+
+    context(@"initWithDictionary:", ^{
+        beforeEach(^{
+            NSDictionary* dict = @{
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLGateStatusSpec.m:47]:
```suggestion
            NSDictionary *dict = @{
```

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLRoofStatusSpec.m [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+#import "SDLRoofStatus.h"
+#import "SDLRPCParameterNames.h"
+#import "SDLWindowState.h"
+
+QuickSpecBegin(SDLRoofStatusSpec)
+
+SDLGrid *location = [[SDLGrid alloc] init];
+SDLDoorStatusType status = SDLDoorStatusTypeAjar;
+SDLWindowState *state = [[SDLWindowState alloc] init];
+__block SDLRoofStatus* testStruct = nil;
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLRoofStatusSpec.m:22]:
```suggestion
__block SDLRoofStatus *testStruct = nil;
```

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLRoofStatusSpec.m [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+describe(@"getter/setter tests", ^{
+    afterEach(^{
+        testStruct = nil;
+    });
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLRoofStatusSpec.m:27]: This `afterEach(^{})` is unnecessary and can be removed.

File: SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLRoofStatusSpec.m [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        it(@"expect all properties to be set correctly", ^{
+            expect(testStruct.location).to(equal(location));
+            expect(testStruct.status).to(equal(status));
+            expect(testStruct.state).to(equal(state));
+        });
+    });
+
+    context(@"initWithDictionary:", ^{
+        beforeEach(^{
+            NSDictionary* dict = @{
```
Comment (NicoleYarroch) [inline SmartDeviceLinkTests/RPCSpecs/StructSpecs/SDLRoofStatusSpec.m:50]:
```suggestion
            NSDictionary *dict = @{
```

**Review by NicoleYarroch — CHANGES_REQUESTED** [review]
Can you change the formatting of the `init`s back to the way it was in the first review? I also requested a few minor changes to the tests.

**Comment by NicoleYarroch** [comment]
@yoooriii Can you tag me when you are ready to have Livio review the PR again? Thanks you.

**Comment by yoooriii** [comment]
Hello @NicoleYarroch. I have updated the PR, could you please proceed ?

**Review by NicoleYarroch — APPROVED** [review]

**Pull request merged**